File: State/DotNet/State/GumballMachine.cs
namespace State;

public class GumballMachine
{
    public State SoldOutState;
    public State NoQuarterState;
    public State HasQuarterState;
    public State SoldState;
    public State WinnerState;
    public State State { get; set; }
    public int Count { get; set; } = 0;

    public GumballMachine(int numberGumballs)
    {
        SoldOutState = new SoldOutState(this);
        NoQuarterState = new NoQuarterState(this);
        HasQuarterState = new HasQuarterState(this);
        SoldState = new SoldState(this);
        WinnerState = new WinnerState(this);

        Count = numberGumballs;
        if (numberGumballs > 0)
        {
            State = NoQuarterState;
        }
        else
        {
            State = SoldOutState;
        }
    }

    public void InsertQuarter()
    {
        State.InsertQuarter();
    }

    public void EjectQuarter()
    {
        State.EjectQuarter();
    }

    public void TurnCrank()
    {
        State.TurnCrank();
        State.Dispense();
    }

    public void Refill(int count)
    {
        Count += count;
        Console.WriteLine(
            $"The gumball machine was just refilled; its new count is: {Count}"
        );
        State.Refill();
    }

    public void ReleaseBall()
    {
        Console.WriteLine("A gumball comes rolling out the slot...");
        if (Count > 0)
        {
            Count -= 1;
        }
    }

    public override string ToString()
    {
        string message = "Mighty Gumball, Inc.\n";
        message += "Standing Gumball Model #2004\n";
        message += $"Inventory: {Count} gumballs\n";
        if (State == HasQuarterState)
        {
            message += "Machine is waiting for quarter";
        }
        else if (State == NoQuarterState)
        {
            message += "Machine is waiting for action: turn crank or eject quarter";
        }
        else if (State == SoldOutState)
        {
            message += "Machine is sold out";
        }
        else if (State == SoldState)
        {
            message += "Machine is dispensing gumball";
        }
        message += "\n";
        return message;
    }
}

File: State/Python/src/GumballMachine.py
class GumballMachine:
    def __init__(self, gumballs_num: int) -> None:
        from . import States

        self.sold_out_state = States.SoldOutState(self)
        self.no_quarter_state = States.NoQuarterState(self)
        self.has_quarter_state = States.HasQuarterState(self)
        self.sold_state = States.SoldState(self)
        self.winner_state = States.WinnerState(self)
        self.count = gumballs_num
        if gumballs_num > 0:
            self.state = self.no_quarter_state
        else:
            self.state = self.sold_out_state

    def __str__(self) -> str:
        string = "Mighty Gumball, Inc.\n"
        string += "Standing Gumball Model #2004\n"
        string += f"Inventory: {self.count} gumballs\n"
        string += f"{str(self.state)}\n"
        return string

    def insert_quarter(self) -> None:
        self.state.insert_quarter()

    def eject_quarter(self) -> None:
        self.state.eject_quarter()

    def turn_crank(self) -> None:
        self.state.turn_crank()
        self.state.dispense()

    def refill(self, count: int) -> None:
        self.count += count
        print(f"The gumball machine was just refilled its new count is: {self.count}")
        self.state.refill()

    def release_ball(self) -> None:
        print("A gumball comes rolling out the slot...")
        if self.count > 0:
            self.count -= 1